File: scala/SimpleAk/src/main/scala-2.11/actor/AkkaDb.scala
package actor

import java.util

import akka.actor.{Status, Actor}
import akka.event.{Logging, LoggingAdapter}
import mol.{KeyNotFoundException, GetRequest, SetRequest}


/**
 * Created by cqshinn on 2/11/16.
 */
class AkkaDb extends Actor{
  val log: LoggingAdapter = Logging(context.system, this)
  val map: util.Map[Option[String],Option[String]] = new util.HashMap[Option[String],Option[String]]

  override def receive = {
    case SetRequest(key,value) => {
      log.info("Received set request – key: {}\n value: {}\n", key, value)
      map.put(key,value)
    }
    case GetRequest(key) => {
      log.info("Received set request – key: {}\n", key)
      map.get(key) match {
        case Some(x) => sender ! x
        case None => sender ! Status.Failure(KeyNotFoundException(key))
      }
    }
    case obj => sender ! Status.Failure(new ClassNotFoundException("not found"))
  }
}

File: java/SimpleAk/src/main/java/actor/AkkaDb.java
package actor;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import com.google.common.collect.Maps;
import mol.GetRequest;
import mol.KeyNotFoundException;
import mol.SetRequest;

import java.util.Map;

/**
 * Created by cqshinn on 2/11/16.
 */
public class AkkaDb extends AbstractActor {
    protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
    public final Map<String, Object> map = Maps.newHashMap();

    private AkkaDb() {
        receive(ReceiveBuilder
                        .match(SetRequest.class, mes -> {
                            log.info("Received set request – key: {}\n value: {}\n", mes.getKey(), mes.getValue());
                            map.put(mes.getKey(), mes.getValue());
                        })
                        .match(GetRequest.class, mes -> {
                            log.info("Received get request – key: {}\n", mes);
                            String value = (String)map.get(mes.key);
                            Object response = value != null ? value : new Status.Failure(new KeyNotFoundException(mes.key));
                            sender().tell(response, self());
                        })
                        .matchAny(obj -> {
                            log.info("receive unknown obj {}\n", obj);
                            sender().tell(new Status.Failure(new ClassNotFoundException("not found obj")), self());
                        })
                        .build()
        );
    }
}
